File: HW1/p1/Histogram.cpp
#include "Histogram.h"
#include <opencv2/opencv.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/plot.hpp>
#include <cmath>

Histogram::Histogram(int h, int w):height(h), width(w), BytesPerPixel(1){

}

// Read image (filename specified by first argument) into image data matrix
void Histogram::readRawImage(std::string path){
    FILE *file;
	
    if (!(file=fopen(path.c_str(), "rb"))) {
		std::cout << "Cannot open file: " << path << std::endl;
		exit(1);
	}

    unsigned char Imagedata[height][width][BytesPerPixel];
	fread(Imagedata, sizeof(unsigned char), width * height * BytesPerPixel, file);

    image.resize(height, std::vector<std::vector<unsigned char>>(width, std::vector<unsigned char>(BytesPerPixel, 0)));
    for(int i = 0; i < height; i++){
        for(int j = 0; j < width; j++){
            image[i][j][0] = Imagedata[i][j][0];
        }
    }
	fclose(file);
}

// Write image data (filename specified by second argument) from image data matrix
void Histogram::writeRawImage(std::string path, int type, std::vector<std::vector<std::vector<unsigned char> > > wImage){
    FILE *file;

	if (!(file=fopen(path.c_str(), "wb"))) {
		std::cout << "Cannot open file: " << path << std::endl;
		exit(1);
	}
    
    if(type == 0){
        unsigned char Imagedata[height][width];
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                Imagedata[i][j] = wImage[i][j][0];
            }
        }
     
        fwrite(Imagedata, sizeof(unsigned char), width * height, file);
    }else if(type == 1){
        unsigned char Imagedata[height][width][3];
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                for(int k = 0; k < 3; k++){
                    Imagedata[i][j][k] = wImage[i][j][k];
                }
            }
        }
        fwrite(Imagedata, sizeof(unsigned char), width * height * 3, file);
    }
    
	fclose(file);
}

void Histogram::writeRawImage(std::string path){
    FILE *file;

	if (!(file=fopen(path.c_str(), "wb"))) {
		std::cout << "Cannot open file: " << path << std::endl;
		exit(1);
	}
    
    if(BytesPerPixel == 1){
        unsigned char Imagedata[height][width];
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                Imagedata[i][j] = image[i][j][0];
            }
        }
     
        fwrite(Imagedata, sizeof(unsigned char), width * height, file);
    }else if(BytesPerPixel == 3){
        unsigned char Imagedata[height][width][3];
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                for(int k = 0; k < 3; k++){
                    Imagedata[i][j][k] = image[i][j][k];
                }
            }
        }
        fwrite(Imagedata, sizeof(unsigned char), width * height * 3, file);
    }
    
	fclose(file);

}

void Histogram::plotOrignalImage(){
    double pixelArray[256] = {0};
    for(int i = 0; i < height; i++){
        for(int j = 0; j < width; j++){
            pixelArray[image[i][j][0]]++;
        }
    }

    cv::Mat dataMat(1, 256, CV_64F, pixelArray);
    cv::Mat plotResult;
    cv::Ptr<cv::plot::Plot2d> plot = cv::plot::Plot2d::create(dataMat);

    plot->setPlotBackgroundColor(cv::Scalar(255, 255, 255));
    plot->setPlotLineColor(cv::Scalar(0, 0, 255));
    plot->setShowText(true);
    plot->setMinX(0);
    plot->setMaxX(256); 
    plot->setMinY(0);  
    plot->setMaxY(1000);

    plot->render(plotResult);
    cv::copyMakeBorder(plotResult, plotResult, 20, 20, 20, 20, cv::BORDER_CONSTANT, cv::Scalar(255, 255, 255));
    cv::namedWindow("Orignal Histogram", cv::WINDOW_AUTOSIZE);
    cv::imshow("Orignal Histogram", plotResult);

    cv::waitKey(0);
}

std::vector<std::vector<std::vector<unsigned char> > > Histogram::transferFunctionBased(){
    std::vector<std::vector<std::vector<unsigned char>>> result(height, std::vector<std::vector<unsigned char>>(width, std::vector<unsigned char>(3, 0)));
    double pixelArray[256] = {0.0};
    // Count the frequency of pixels of each grayscale value
    for(int i = 0; i < height; i++){
        for(int j = 0; j < width; j++){
            pixelArray[image[i][j][0]]++;
        }
    }

    // Calculate the normalized probability histogram
    double totalPixel = width * height;
    for(int i = 0; i < 256; i++){
        pixelArray[i] /= totalPixel;
    }

    // Calculate the CDF
    double accumulate = 0.0;
    for(int i = 0; i < 256; i++){
        accumulate += pixelArray[i];
        pixelArray[i] = accumulate;
    }

    // Create the mapping-table
    int mappingTable[256] = {0};
    for(int i = 0; i < 256; i++){
        mappingTable[i] = static_cast<int>(floor(pixelArray[i] * 255));
        if(mappingTable[i] > 255){
            mappingTable[i] = 255;
        }
    }

    // mapping to image
    for(int i = 0; i < height; i++){
        for(int j = 0; j < width; j++){
            result[i][j][0] = static_cast<unsigned char>(mappingTable[image[i][j][0]]);
        }
    }
    
    return result;
}

std::vector<std::vector<std::vector<unsigned char> > > Histogram::bucketFilling(){
    std::vector<std::vector<std::vector<unsigned char>>> result(height, std::vector<std::vector<unsigned char>>(width, std::vector<unsigned char>(3, 0)));
    int pixelArray[256] = {0};
    // Count the frequency of pixels of each grayscale value
    for(int i = 0; i < height; i++){
        for(int j = 0; j < width; j++){
            pixelArray[image[i][j][0]]++;
        }
    }

    int accumulate = 0;
    int accPixel[256] = {0};
    for(int i = 0; i < 256; i++){
        accPixel[i] = accumulate;
        accumulate += pixelArray[i];
    }

    int binNumber = height * width / 256;
    int pixelNumber[256] = {0};
    
    // mapping to image
    for(int i = 0; i < height; i++){
        for(int j = 0; j < width; j++){
            pixelNumber[image[i][j][0]]++;
            int mapping = (accPixel[image[i][j][0]] + pixelNumber[image[i][j][0]]) / binNumber;
            if(mapping > 255){
                mapping = 255;
            }
            result[i][j][0] = mapping;
        }
    }

    return result;

}

File: HW1/p1/Histogram.py
import numpy as np
import matplotlib.pyplot as plt
import math

def read_raw_image(file_path, height, width):
    # read raw image
    with open(file_path, 'rb') as file:
        raw_data = np.fromfile(file, dtype=np.uint8)
    
    # convert shape
    image = raw_data.reshape((height, width))

    return image

def plot_histogram(image):
    # flat image
    flattened_image = image.flatten()

    # plot histogram
    y_values = [0] * 256
    for pixel_value in flattened_image:
        y_values[pixel_value] += 1
    
    x_values = list(range(256))
    plt.bar(x_values, y_values, color='blue')
    plt.title('Histogram of Original Image')
    plt.xlabel('Pixel Intensity')
    plt.ylabel('Number of Pixels')
    plt.xlim(0, 256)
    plt.ylim(0, 10000)
    plt.show()

def transfer_function_based(image):
    # flat image
    flattened_image = image.flatten()

    # Count the frequency of pixels of each grayscale value
    y_values = [0] * 256
    for pixel_value in flattened_image:
        y_values[pixel_value] += 1
    x_values = list(y_values)
    
    # Calculate the normalized probability histogram
    for i in range(256):
        x_values[i] /= len(flattened_image)
    
    # Calculate the CDF
    accumulate = 0.0
    for i in range(256):
        accumulate += x_values[i]
        x_values[i] = accumulate
    
    #  Create the mapping-table
    for i in range(256):
        x_values[i] = math.floor(x_values[i] * 255)

    # plot the histogram
    plt.bar(x_values, y_values, color='blue')
    plt.title('Transfer Function Based Histogram')
    plt.xlabel('Pixel Intensity')
    plt.ylabel('Number of Pixels')
    plt.xlim(0, 256)
    plt.ylim(0, 3000)
    plt.show()
    
def bucket_filling(image):
    # flat image
    flattened_image = image.flatten()

    y_values = [0] * 256
    x_values = list(range(256))
    totolPixels = len(flattened_image)
    binNumber = math.floor(totolPixels / 256)
    restPixel = totolPixels % 256
    for i in range(256):
        y_values[i] = binNumber
    for i in range(restPixel):
        y_values[i] += 1
    
    # plot the histogram
    plt.bar(x_values, y_values, color='blue')
    plt.title('Bucket Filling Histogram')
    plt.xlabel('Pixel Intensity')
    plt.ylabel('Number of Pixels')
    plt.xlim(0, 256)
    plt.ylim(0, 1200)
    plt.show()

    



raw_file_path = '/Users/ren/Documents/hw/EE569/HW1_materials/images/DimLight.raw'
image_height = 340
image_width = 596

image = read_raw_image(raw_file_path, image_height, image_width)
plot_histogram(image)
transfer_function_based(image)
bucket_filling(image)
